README.rst
========================================================
EMDB-SFF Read/Write Toolkit (``sfftk-rw``)
========================================================

.. image:: https://img.shields.io/pypi/v/sfftk-rw?color=green
    :alt: PyPI

.. image:: https://img.shields.io/pypi/pyversions/sfftk-rw
    :alt: PyPI - Python Version

.. image:: https://travis-ci.org/emdb-empiar/sfftk-rw.svg?branch=master
    :target: https://travis-ci.org/emdb-empiar/sfftk-rw

.. image:: https://readthedocs.org/projects/sfftk-rw/badge/?version=latest
    :target: https://sfftk-rw.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://codecov.io/gh/emdb-empiar/sfftk-rw/branch/master/graph/badge.svg?token=OIN3VDFH1K
    :target: https://codecov.io/gh/emdb-empiar/sfftk-rw



-------------
About
-------------

``sfftk-rw`` is a Python toolkit for *reading and writing EMDB-SFF files only*. It is part of a family of tools
designed to work with EMDB-SFF files. Other related tools are:

-   ``sfftk`` - depends on ``sfftk-rw`` and additionally:

    *   **converts** application-specific segmentation files to valid EMDB-SFF files;

    *   **annotates** EMDB-SFF files;

    *   **prepares** segmentation files prior to conversion to EMDB-SFF files;

-   **SAT** - an online wrapper around ``sfftk`` available at https://www.ebi.ac.uk/empiar/sat/

-------------
License
-------------

``sfftk-rw`` is free and open source software released under the terms of the Apache License,
Version 2.0. Source code is copyright **EMBL-European Bioinformatics Institute (EMBL-EBI) 2017**.



CHANGELOG.md
#Changes by release

## [0.8.1] - 2023-09-26

* bugfix for using `SubtomogramAverage` class in `SFFShapePrimitiveList` class

## [0.8.0] - 2023-09-25

* added `SFFSubtomogramAverage` class

## [0.7.3] - 2023-01-27 Option to include source EMDB-SFF colours when merging

* In EMDB-SFF `v0.8.0.dev1` `SFFSegmentation.merge_annotation(other_seg, include_colour=False)` (*do not* merge colour by default)

## [0.7.2] - 2022-04-26

add Python3.10 support

* removed use of `os` package for exit codes (problems on Windows)
* fixed errors in `developing` documentation

## [0.7.1] - 2021-08-24

bugfix: reading from HDF5 retains some ids as `numpy.int64`

* tests added and pass

## [0.7.0] - 2021-06-17

add Python3.9 support; dropped Python2.7

* add official support for Python3.9
* dropped support for Python2.7 and Python3.5
* now in beta but stable
* new version: v0.7.0

## [0.6.7.dev1] - 2020-11-10

dropped py39 until time is available ;-)
version bump: v0.6.7.dev1

## [0.6.7.dev0] - 2020-11-09

h5py accommomdation

* h5py>3 has dropped py27 support; it now returns strings as bytes
* this release fixes reading of bytes as unicode strings
* also works with py39
* modified h5py version for py35
* version bump: v0.6.7.dev0

## [0.6.6.dev0] - 2020-05-15

Bugfix: saving excluding name, software and details

* merge_annotation method had these fields commented out; fixed
* unit tests updated to check the same
* version bump: v0.6.6.dev0

## [0.6.5.dev0] - 2020-05-15

Bugfix: index collision during append for new objects to lists

* during annotation in `sfftk` new objects e.g. `SFFSoftware` or `SFFExternalReference` objects, are
created with an `id=0`; an exception is raised during addition to the `SFFListType._id_dict` if the ID
already exists;
* fix: a new method `_get_next_id` has been added to `SFFListType` which determines the next best ID
to use to avoid collision; this also works when items with `id=None` are added - a new ID (starting from 1)
is issued to ensure the integrity of `SFFListType._id_dict`
* version bump: v0.6.5.dev0

## [0.6.4.dev0] - 2020-04-15

Bugfix: reading large XML files

* version bump: v0.6.4.dev0

## [0.6.3.dev0] - 2020-04-14

Bugfix: reading of empty bounding box in v0.7.0.dev0 files

* version bump: v0.6.3.dev0

## [0.6.2.dev6] - 2020-03-30

Development changes

* removed redudant `-t/--top-level-only` convert flag

## [0.6.2.dev5] - 2020-03-26

Development changes

* minor documentation corrections
* handling `MemoryError` when encoding large 3D volumes for travis
* version bump: v0.6.2.dev5

## [0.6.2.dev4] - 2020-08-18

Development bugfix

* utils.py: `get_path()` now returns `None` if path invalid
* version bump: v0.6.2.dev4

## [0.6.2.dev3] - 2020-08-18

Development bugfix

* `sfftk` will now use core tools from `sfftk-rw`
* use `iter*` methods from `_dict` (dictionary preserving insertion order) instead of from builtin dictionary
* version bump: v0.6.2.dev3

## [0.6.2.dev2] - 2020-03-13

Development changes

* view parser: check EMDB-SFF version with `--sff-version`
* handling of IMOD chunks deferred to `sfftk` package
* this change has no impact whatsover on function of the package

## [0.6.2.dev1] - 2020-03-13

Development changes

* convert parser argument `from_file` has `nargs='*'` for `sfftk` downstream
* this change has no impact whatsover on function of the package

## [0.6.2.dev0] - 2020-03-06

added methods for handling notes to `sfftkrw.SFFSegmentation`

## [0.6.1.dev0] - 2020-03-02

Bugfix and minor improvements

Fixed a bug which affected encoded data (lattice, vertices, normals and triangles).

* reverted presentation of `SFFVolume` subclasses to be ordered as (rows, cols, sections)
* now we store encoded data as unicode strings instead of bytes; otherwise, on export we write string with `'b"..."` formatting
* writing to JSON and HDF5 is not raw bytes but Unicode string
* version bump: v0.6.1.dev0


## [0.6.0.dev2] - 2020-02-25

* `-x/--exclude-geometry` for JSON output
* `--json-sort` and `--json-indent <+int>` to control JSON output
* now links to h5py documentation should work
* changed interface for v0.7.0 adapter HDF5 methods in line with sfftk-rw v0.6.0.dev0 changes
* now `args` passed through all classes for command-line contol/features

## [0.6.0.dev1] - 2020-02-24

* clean up of documentation to include correct examples, adapters and generateDS API
* now `from sfftkrw import SFFSegmentation` or `import sfftkrw as sff`
* new images for data model page
* new module `sfftkrw/conf.py` for settings; removed `sfftkrw/api.py`
* added new data files

## [0.6.0.dev0] - 2020-02-21

This is the first release to work with EMDB-SFF v0.8.0 files in addition to v0.7 files. When reading 
the API to use is determined from the version attribute. However, creating new files will only
be of v0.8.0.

* new `sfftkrw.api` provided a short path to v0.8 adapter e.g. `sfftkrw.api.SFFSegmentation`
* documentation improvements
* changed entry point to `sff` (from `sff-rw`)
* added references for Python version compatibility: `_FileNotFound`, `_classic_dict` 
* added missing field `attribute` to shapes
* changed `SFFSegmentation` container attributes to `*_list` e.g. `segment_list` instead of `segments`. 
However we retain `segments` as a property for backward compatibility.
* signature of HDF5 conversion methods (`as_hff` and `from_hff`) is now `*(parent_group, args, name=None)`
by default
* refactored `SFFSegmentation.from_file`
* new `SFFSegmentation.to_file` which is an alias for `SFFSegmentation.export`
* added test data under v0.8 schema

## [0.5.2.dev1] - 2019-12-18

### List validations

* schema/base.py: `SFFListType` uses `min_length` attribute to check that minimum length constraint is satisfied
* schema/adapter.py: `min_length` attribute set for `SFFPolygonList` (1) and `SFFVertexList` (3)
* added unit test for the above
* deleted config stub file
* version bump: v0.5.2.dev1

## [0.5.2.dev0] - 2019-12-17

### Major improvements in handling of HDF5 and JSON


sfftkrw/core/parser.py
# -*- coding: utf-8 -*-
# parser.py
"""
parser.py
=========


"""
from __future__ import print_function

import argparse
import os
import re

from . import _dict_iter_keys
from .print_tools import print_date
from .. import SFFTKRW_VERSION, SFFTKRW_ENTRY_POINT, SUPPORTED_EMDB_SFF_VERSIONS
from ..core import _decode, _basestring

__author__ = 'Paul K. Korir, PhD'
__email__ = 'pkorir@ebi.ac.uk, paul.korir@gmail.com'
__date__ = '2016-06-10'
__updated__ = '2018-02-14'

verbosity_range = list(range(4))
multi_file_formats = ['stl', 'map', 'mrc', 'rec']
prepable_file_formats = ['mrc', 'map', 'rec']
rescalable_file_formats = ['stl']


def add_args(parser, the_arg):
    """Convenience function to add ``the_arg`` to the ``parser``.

    This relies on the argument being structured as a dictionary with the keys 
    ``args`` for positional arguments and ``kwargs`` for the keyword
    arguments. The value of doing this is that arguments that are reused
    in several parsers can be referred to by a variable instead of being 
    redefined. 

    Usage::

        >>> my_arg = {'arg': ['-x'], 'kwargs': {'help': 'help'}}
        >>> this_parser = argparse.ArgumentParser()
        >>> add_args(this_parser, my_arg)

    :param parser: the parser to be modified
    :type parser: :py:class:`argparse.ArgumentParser`
    :param dict the_arg: the argument specified as a dict with keys ``args`` and ``kwargs``
    :return: a modified parser object
    :rtype: :py:class:`argparse.ArgumentParser`
    """
    return parser.add_argument(*the_arg['args'], **the_arg['kwargs'])


Parser = argparse.ArgumentParser(
    prog=SFFTKRW_ENTRY_POINT, description="The EMDB-SFF Read/Write Toolkit (sfftk-rw)")
Parser.add_argument(
    '-V', '--version',
    action='store_true',
    default=False,
    help='show the sfftk-rw version string and the supported EMDB-SFF Read/Write version string',
)

subparsers = Parser.add_subparsers(
    title='Tools',
    dest='subcommand',
    description='The EMDB-SFF Read/Write Toolkit ({}) provides the following tools:'.format(SFFTKRW_ENTRY_POINT),
    metavar="EMDB-SFF Read/Write Tools"
)

# =========================================================================
# common arguments
# =========================================================================
sff_file = {
    'args': ['sff_file'],
    'kwargs': {
        'help': 'path (rel/abs) to an EMDB-SFF file',
    }
}
details = {
    'args': ['-D', '--details'],
    'kwargs': {
        'help': "populates <details>...</details> in the XML file"
    }
}

FORMAT_LIST = [
    ('sff', 'XML'),
    ('hff', 'HDF5'),
    ('json', 'JSON'),
]
format_ = {
    'args': ['-f', '--format'],
    'kwargs': {
        # 'default': 'sff',
        'help': "output file format; valid options are: {formats} [default: {default}]".format(
            default=FORMAT_LIST[0][0],
            formats=", ".join(map(lambda x: "{} ({})".format(x[0], x[1]), FORMAT_LIST))
        ),
    }
}
exclude_geometry = {
    'args': ['-x', '--exclude-geometry'],
    'kwargs': {
        'default': False,
        'action': 'store_true',
        'help': 'do not include the geometry in the conversion; geometry is included by default [default: False]',
    }
}
header = {
    'args': ['-H', '--header'],
    'kwargs': {
        'default': False,
        'action': 'store_true',
        'help': 'show EMDB-SFF header (global) attributes [default: False]'
    }
}
segment_id = {
    'args': ['-i', '--segment-id'],
    'kwargs': {
        'help': 'refer to a segment by its ID'
    }
}
name = {
    'args': ['-N', '--name'],
    'kwargs': {
        'help': "the segmentation name"
    }
}
output = {
    'args': ['-o', '--output'],
    'kwargs': {
        'default': None,
        'help': "file to convert to; the extension (.sff, .hff, .json) determines the output format [default: None]"
    }
}
primary_descriptor = {
    'args': ['-R', '--primary-descriptor'],
    'kwargs': {
        'help': "populates the <primary_descriptor>...</primary_descriptor> to this value [valid values:  three_d_volume, mesh_list, shape_primitive_list]"
    }
}
segment_name = {
    'args': ['-s', '--segment-name'],
    'kwargs': {
        'help': "the name of the segment"
    }
}
json_sort = {
    'args': ['--json-sort'],
    'kwargs': {
        'default': False,
        'action': 'store_true',
        'help': "output JSON sorted lexicographically [default: False]"
    }
}
json_indent = {
    'args': ['--json-indent'],
    'kwargs': {
        'type': int,
        'default': 2,
        'help': "size in spaces of the JSON indent [default: 2]"
    }
}
verbose = {
    'args': ['-v', '--verbose'],
    'kwargs': {
        'action': 'store_true',
        'default': False,
        'help': "verbose output"
    },
}

# =========================================================================
# convert subparser
# =========================================================================
convert_parser = subparsers.add_parser(
    'convert', description="Perform EMDB-SFF file format interconversions", help="converts between EMDB-SFF formats")
convert_parser.add_argument('from_file', nargs='*', help="file to convert from")
convert_parser.add_argument(*details['args'], **details['kwargs'])
convert_parser.add_argument(
    *primary_descriptor['args'], **primary_descriptor['kwargs'])
convert_parser.add_argument(*verbose['args'], **verbose['kwargs'])
add_args(convert_parser, exclude_geometry)
add_args(convert_parser, json_indent)
add_args(convert_parser, json_sort)
group = convert_parser.add_mutually_exclusive_group()
group.add_argument(*output['args'], **output['kwargs'])
group.add_argument(*format_['args'], **format_['kwargs'])

# =========================================================================
# view subparser
# =========================================================================
view_parser = subparsers.add_parser(
    'view', description="View a summary of an SFF file", help="view file summary")
view_parser.add_argument('from_file', help="any SFF file")
view_parser.add_argument(
    '--sff-version', action='store_true', help="show SFF format version")
view_parser.add_argument(*verbose['args'], **verbose['kwargs'])

# get the full list of tools from the Parser object
tool_list = ['all', 'core', 'schema', 'main']

# tests
test_help = "one or none of the following: {}".format(", ".join(tool_list))
tests_parser = subparsers.add_parser(
    'tests', description="Run unit tests", help="run unit tests")
tests_parser.add_argument('tool', nargs='+', help=test_help)
tests_parser.add_argument('-v', '--verbosity', default=1, type=int,
                          help="set verbosity; valid values: %s [default: 0]" % ", ".join(map(str, verbosity_range)))

tests_parser.add_argument('--dry-run', default=False, action='store_true',
                          help='do not run tests [default: False]')


# parser function
def parse_args(_args, use_shlex=False):
    """
    Parse and check command-line arguments.

    This function does all the heavy lifting in ensuring that commandline
    arguments are properly formatted and checked for sanity.

    In this way command handlers (defined in ``sfftk/sff.py`` e.g. ``handle_convert(...)``)
    assume correct argument values and can concentrate on functionality.

    :param list _args: list of arguments (``use_shlex=False``); string of arguments (``use_shlex=True``)
    :type _args: list or str or unicode
    :param bool use_shlex: treat ``_args`` as a string instead for parsing using ``shlex`` lib
    :return: parsed arguments
    :rtype: :py:class:`argparse.Namespace`
    """
    if use_shlex:  # if we treat _args as a command string for shlex to process
        try:
            assert isinstance(_args, _basestring)
        except AssertionError:
            return 64
        import shlex
        _args = shlex.split(_args)
    # if we have no subcommands then show the available tools
    if len(_args) == 0:
        Parser.print_help()
        return 0
    # if we only have a subcommand then show that subcommand's help
    elif len(_args) == 1:
        # print(_args[0])
        # _print(Parser._actions)
        # if _args[0] == 'tests':
        #     pass
        if _args[0] == '-V' or _args[0] == '--version':
            print_date(
                "sfftk-rw version: {} for EMDB-SFF {}".format(SFFTKRW_VERSION, ', '.join(SUPPORTED_EMDB_SFF_VERSIONS)))
            return 0
        # anytime a new argument is added to the base parser subparsers are bumped down in index
        elif _args[0] in _dict_iter_keys(Parser._actions[2].choices):
            exec('{}_parser.print_help()'.format(_args[0]))
            return 0
    # parse arguments
    args = Parser.parse_args(_args)

    # check values
    # view
    if args.subcommand == 'view':
        pass  # no view-specific checks yet
    # convert
    elif args.subcommand == 'convert':
        # we only use the first file in sfftk-rw; sfftk may use more than one file
        args.from_file = args.from_file[0]
        # convert details to unicode
        if args.details is not None:
            args.details = _decode(args.details, 'utf-8')
        # set the output file
        if args.output is None:
            from_file = args.from_file
            dirname = os.path.dirname(from_file)
            if args.format:
                try:
                    assert args.format in list(map(lambda x: x[0], FORMAT_LIST))
                except AssertionError:
                    print_date("Invalid output format: {invalid}; valid values are: {formats}".format(
                        invalid=args.format,
                        formats=", ".join(map(lambda x: x[0], FORMAT_LIST)))
                    )
                    return 64
                fn = ".".join(os.path.basename(from_file).split(
                    '.')[:-1]) + '.{}'.format(args.format)
                args.__setattr__('output', os.path.join(dirname, fn))
            # convert file.sff to file.hff
            elif re.match(r'.*\.(sff|xml)$', from_file):
                fn = ".".join(
                    os.path.basename(from_file).split('.')[:-1]) + '.hff'
                args.__setattr__('output', os.path.join(dirname, fn))
            # convert file.hff to file.sff
            elif re.match(r'.*\.(hff|h5|hdf5)$', from_file):
                fn = ".".join(
                    os.path.basename(from_file).split('.')[:-1]) + '.sff'
                args.__setattr__('output', os.path.join(dirname, fn))
            else:
                fn = ".".join(
                    os.path.basename(from_file).split('.')[:-1]) + '.sff'
                args.__setattr__('output', os.path.join(dirname, fn))
            if args.verbose:
                print_date("Setting output file to {}".format(args.output))

        # ensure valid primary_descriptor
        if args.primary_descriptor:
            try:
                assert args.primary_descriptor in [
                    u'threeDVolume', u'meshList', u'shapePrimitiveList',
                    u'three_d_volume', u'mesh_list', u'shape_primitive_list',
                ]
            except AssertionError:
                print_date(
                    "Invalid value for primary descriptor: {}".format(args.primary_descriptor))
                return 64
            if args.verbose:
                print_date(
                    "Trying to set primary descriptor to {}".format(args.primary_descriptor))

        # report on geometry
        if args.exclude_geometry and (args.format == u'json' or re.match(r'.*\.(json)$', args.output, re.IGNORECASE)):
            print_date("Excluding geometry for JSON")

        # validate indent for json
        if args.format == u'json' or re.match(r'.*\.(json)$', args.output, re.IGNORECASE):
            try:
                assert args.json_indent >= 0
            except AssertionError:
                print_date("Invalid value for --json-indent: {}".format(args.json_indent))
                return 64
            if args.verbose:
                print_date("Indenting JSON with indent={}".format(args.json_indent))

            if args.json_sort and args.verbose:
                print_date("JSON keys will be sorted lexicographically")

    # tests
    elif args.subcommand == 'tests':
        # normalise tool list
        # if 'all' is specified together with others then it should simply be 'all'
        if 'all' in args.tool:
            args.tool = ['all']
        # if isinstance(args.tool, list):
        for tool in args.tool:
            try:
                assert tool in tool_list
            except AssertionError:
                print_date(
                    "Unknown tool: {}; Available tools for test: {}".format(tool, ", ".join(tool_list))
                )
                return 64
        if args.verbosity:
            try:
                assert args.verbosity in range(4)
            except:
                print_date(
                    "Verbosity should be in {}-{}: {} given".format(
                        verbosity_range[0],
                        verbosity_range[-1],
                        args.verbosity
                    )
                )
                return 64

    return args
